lib/base/io-engine.cpp

```diff
@@ -16,60 +16,87 @@
+			if (ie.m_CpuBoundSemaphore.compare_exchange_weak(
+				freeSlots,
+				freeSlots - 1,
+				std::memory_order_seq_cst, // The default memory order in case of success
+				loadOrder // A failure here is equivalent to loading the latest value into freeSlots
+			)) {
+				return true;
+			}
+		}
 
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
+		return false;
+	};
+
+	while (!fastPath(std::memory_order_relaxed)) {
+		if (!cv) {
+			cv = Shared<AsioConditionVariable>::Make(ie.GetIoContext());
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
+			// The above line may take a little bit, so let's optimistically re-check
+			if (fastPath(std::memory_order_relaxed)) {
+				break;
+			}
 		}
 
-		break;
-	}
-}
+		{
+			std::unique_lock lock (ie.m_CpuBoundWaitingMutex);
 
-CpuBoundWork::~CpuBoundWork()
-{
-	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
+			// The above line may take even longer, so let's check again.
+			// Also mitigate lost wake-ups by synchronizing with Done().
+			// On details, see Done() implementation inline comments.
+			if (fastPath()) {
+				break;
+			}
+
+			ie.m_CpuBoundWaiting.emplace_back(strand, cv);
+		}
+
+		cv->Wait(yc);
 	}
 }
 
 void CpuBoundWork::Done()
 {
 	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
-
 		m_Done = true;
-	}
-}
-
-IoBoundWorkSlot::IoBoundWorkSlot(boost::asio::yield_context yc)
-	: yc(yc)
-{
-	IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
-}
-
-IoBoundWorkSlot::~IoBoundWorkSlot()
-{
-	auto& ioEngine (IoEngine::Get());
-
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
+		auto& ie (IoEngine::Get());
+
+		// The constructor takes the slow path only if the semaphore is full,
+		// so we only have to wake up constructors if the semaphore was full.
+		// This works because after fetch_add(), fast paths in the constructor will succeed.
+		if (ie.m_CpuBoundSemaphore.fetch_add(1) < 1) {
+			// So now there are only slow path subscribers from just before the fetch_add() to be woken up.
+			// Precisely, only subscribers from just before the fetch_add() which turned 0 to 1.
+
+			decltype(ie.m_CpuBoundWaiting) subscribers;
+
+			{
+				// Due to the mutex, wake-ups obviously may happen only before or after constructor subscribtions.
+				// Ideally, this wake-up happens after any ongoing subscription (obviously).
+				std::unique_lock lock (ie.m_CpuBoundWaitingMutex);
+				std::swap(subscribers, ie.m_CpuBoundWaiting);
+			}
+			// But even if it happens before (and gets lost), it's not a problem because now the ongoing subscriber
+			// will lock the mutex and re-check the semaphore which is already >0 (fast path) due to fetch_add().
+
+			for (auto& subscriber : subscribers) {
```

Comment:
Unpacking the pair should make the code a bit nicer:

```suggestion
			for (auto& [strand, cv] : subscribers) {
```